src/afMessageLabel.js

```diff
@@ -1,24 +1,39 @@
 angular.module('angularFormMessages')
   .directive('afMessageLabel', function (
+    $interpolate,
     $log,
-    MessageService,
-    translateFilter,
-    TranslateService
+    $translate,
+    AfMessageService
   ) {
     return {
       restrict: 'A',
-      require: '^afMessage',
-      link: function ($scope, elem, attrs, afMessageCtrl) {
+      require: ['^form', '^afMessages'],
+      link: function ($scope, elem, attrs, ctrls) {
         attrs.$observe('afMessageLabel', function (newVal) {
-          var
-            specificLabel = afMessageCtrl.messageId + '.' + newVal,
-            genericLabel = MessageService.getGenericLabelPrefix() + newVal,
-            translation = TranslateService.hasLabel(specificLabel) ? translateFilter(specificLabel) : translateFilter(genericLabel);
+          function translate(translation) {
+            elem.html(translation);
+          }
 
-          if (translation === undefined) {
-            $log.warn('Missing label: \'' + specificLabel + '\' (specific) or \'' + genericLabel + '\' (generic)');
+          if (!newVal) {
+            return;
           }
-          elem.html(translation === undefined ? newVal : translation);
+
+          var
+            formCtrl = ctrls[0],
+            afMessagesCtrl = ctrls[1],
+            specificLabel = $interpolate(formCtrl.$name)($scope) + '.' + (afMessagesCtrl.messageId || afMessagesCtrl.messageIdPrefix) + '.' + newVal,
+            genericLabel = AfMessageService.getGenericLabelPrefix() + newVal;
+
+          $translate(specificLabel)
+            .then(translate)
+            .catch(function () {
+              $translate(genericLabel)
+                .then(translate)
+                .catch(function () {
```

Comment:
Lijkt een beetje een pyramid of doom... Maar ik zie ook geen manier om dat mooier op te lossen